Simplify callfunc more more

smuzaffar opened the pull request.

guitargeek committed:
Avoid using deperecated `RooCatType` in RooStats tutorials

With the rewrite of the RooCategory classes, the `RooCatType` was
deprecated (moved into RooFitLegacy) and should also not be used anymore
in tutorial code.

ferdymercury committed:
Use safer C++ casting

eguiraud committed:
[DF] Teach RColumnRegister to create and store col. readers

With this improvement RColumnRegister can act as a central hub
that creates RDefineReaders and RVariationReaders as needed and
distributes them to the nodes in the computation graph.

eguiraud committed:
[DF] Centralize RDefineReaders and RVariationReaders

Rather than constructing a different reader for each node that
needs it, with this patch we now leverage RColumnRegister as
a "cache" of RDefineReaders and RVariationReaders so that nodes
in the computation graph share the same reader objects (per
processing slot). This is analogous to what RLoopManager does
for RDSColumnReaders and RTreeColumnReaders.

Sharing column readers across the computation graph will be useful
for bulk loading of event values (which we do not want to repeat
for different instances of a given column's reader).

vgvassilev committed:
Simplify InitRetAndExecBuiltin

vgvassilev committed:
Simplify InitRetAndExecNoCtor

vgvassilev committed:
Inline InitRetAndExecBuiltin

vgvassilev committed:
Move the error checks before the lock.

vgvassilev committed:
Inline InitRetAndExecNoCtor

vgvassilev committed:
Take into account the implicit this

alja committed:
Add visualization of hexagons as a type of REveBoxSet

vepadulano committed:
[DF] Catch and propagate C++ exceptions in the distributed tasks

C++ exceptions are instances of some class derived from std::exception.
This cannot be serialized by ROOT because it's missing a dictionary or
streamer. Dask serializes Python exception objects in order to return
them to the user side. So when a C++ exception happens in a Dask app,
the following kind of errors appear:

Warning in <TBufferFile::WriteObjectAny>: since runtime_error has no public constructor
	which can be called without argument, objects of this class
	can not be read with the current library. You will need to
	add a default constructor before attempting to read it.
Warning in <TStreamerInfo::Build>: runtime_error: base class exception has no streamer or dictionary it will not be saved
Warning in <TStreamerInfo::Build>: runtime_error: __cow_string has no streamer or dictionary, data member "_M_msg" will not be saved
Error in <TClass::New>: cannot create object of class runtime_error
Error in <TBufferFile::ReadObject>: could not create object of class runtime_error
 *** Break *** bus error

This commit catches all C++ exceptions in the task, before its end.
Then, the content of the exception is wrapped in a Python RuntimeError
that Dask can properly serialize and return to the user.

Spark is not affected by this issue.

junaire committed:
Drop llvm::GlobalValue in DeferredDeclsToEmit

The previous cling patches wraps `DeferredDeclsToEmit(vector<GlobalDecl>)`
into a struct that contains a `llvm::GlobalVlaue`, but it seems that this
field is useless and can be droped. I have tested this patch in Cling and
spotted no new tests failing, so let's give ROOT a try.

Signed-off-by: Jun Zhang <jun@junz.org>

ikabadzhov committed:
[DF] Add a missing include in df103

ikabadzhov committed:
[DF] Discard warnings in df103

The produced TCanvas has no name now.
This was already the behaviour in the python tutorial.

ikabadzhov committed:
[DF] Remove unnecessary brackets in Defines in df103

alja committed:
Use FlexBox instead of Splitter

eguiraud committed:
[DF] Fix jitted expressions with sub-branches of aliases

Given a string expression such as "alias.subbranch" (where
`subbranch` is _not_ also the name of a valid data member of
the type of the "alias" top-level branch), we used to transform
the expression to `[](T &var0) { return var0.subbranch; }`,
which does not compile.

Now aliases in jitted expressions are resolved in a first step
and only then we try to match the expression against known branch
names, fixing the problem.

This fixes https://github.com/root-project/root/issues/11207 .

eguiraud committed:
[DF] Add test for #11207

linev committed:
Update nlohmann/json.hpp to version 3.10.5

Provide json_fwd.h which can be used now in eve7

linev committed:
[eve7] use nlohmann/json forward declaration from 3.10.5

After upgrading of nlohmann this part also should be upgraded

guitargeek committed:
Fix the classic double "the the" typos in the ROOT documentation

vgvassilev committed:
Add test for ROOT-6523.

vgvassilev committed:
Rely on the underlying type representation to dispatch to the right call.

vgvassilev committed:
Implement more robust conversion via getAs in cling::Value.

Here we enumerate most of the builtin types and we the generate template
specializations for all of them which are capable to perform the correct
conversions.

guitargeek committed:
[RF] Refactor CUDA helpers in RooFitDriver to free functions

This is done to better understand when a function changes the state of
the RooFitDriver.

guitargeek committed:
[RF] Ignore empty spans when filling server values in computeBatch

In `RooAbsReal::computeBatch()`, all server values are filled from the
dataMap. However, it can happen that some values are not in the dataMap
if the server is not a value or shape server. These servers should be
skipped.

guitargeek committed:
[RF] Do initial pass in RooFitDriver::getValHeterogeneous() on GPU

So far, the first pass to measure the timing was done on the CPU and the
second on the GPU. This is not practical when only doing one evaluation
for development purposes. Hence the order is now reversed.

vepadulano committed:
[DF][NFC] Add modules to py_compile sources

vepadulano committed:
[DF] Postpone evaluation of type hints

This commit activates the features of PEP 563 in all modules of DistRDF. This
notably postpones the evaluation of type hints, so that they are not done at
function definition time but only restricted to type checkers. This avoids
issues with circular imports and also helps avoiding unexpected imports. For
example, the import of the `Proxy` module in the `__init__.py` file was
done for type hints purposes only, but triggered an unexpected `import ROOT`.

Together with the postponed evaluation, some modules have got a few adjustments
to their type hints.

eguiraud committed:
[NFC] Make nbviewer/SWAN badges more prominent in the doxygen docs

vepadulano committed:
[DF][NFC] Test import DistRDF does not import ROOT

couet committed:
[skip-ci] improve error message (#11235)

bellenot committed:
Disable brotli support feature in Freetype (not only on Apple) (#11242)

* Disable brotli support feature in Freetype (not only on Apple)

Should fix #11239

* Remove code instead of commenting it out

eguiraud committed:
[DF] Re-add a useful transitive include

RDFHelpers.hxx provides ROOT::RDF::RunGraphs(std::vector<RResultHandle>).
In order for the handy implicit conversion from RResultPtr<T> to
RResultHandle to be available for client code that include
RDFHelpers.hxx, it is not enough that this header forward-declares
RResultHandle, but it needs to transitively provide the whole type
declaration.

In practice, before 8bbd543f3d56, this worked fine:

```
...
RunGraphs({resultptr1, resultptr2});
```

while after that commit it would require an extra `#include
<ROOT/RResultHandle.hxx>`. This patch restores the previous situation.

Commit:
Move RSlotStack from RDF to Imt

eguiraud committed:
[DF] Fix potential use-after-delete in sample callbacks

Before this commit, if an action with a sample callback (currently
only Snapshot) or a DefinePerSample went out of scope, we did not
deregister the corresponding callbacks from the RLoopManager, which
would try to run them anyway during the following event loop.
As callbacks typically perform a call on the original action or
define objects, this could cause a use-after-delete.

With this patch, we now associate each callback to the address of
the corresponding node of the computation graph and remove callbacks
when the nodes go out of scope.

This fixes #11222.

eguiraud committed:
[DF] Add regression test for #11222

vgvassilev committed:
Rely on the cling::Value representation rather than a custom one.

Instead of constructing a void* array with arguments given a cling::Value we
should use it's representation and ask for the void* via the getPtr() interface.
In the past we did not have the cling::Value abstraction and we created some
interfaces which reformat the raw memory. However, the implementation of
cling::Value already assumes that the internal representation is formatted
already accordingly.

This patch takes that shortcut and relies on the assumptions already made for
cling::Value.

vgvassilev committed:
Inline InitRetAndExec

vgvassilev committed:
Further simplify by checking for GetValAddr() early.

eguiraud committed:
[DF] Fix builds with imt=OFF

A recent commit moved RSlotStack to core/imt, so now we can only
include it if R__USE_IMT is defined.

hageboeck committed:
[tree] Reduce string allocations in TBranch::GetFullName.

When called repeatedly via TTree::GetBranch, the string allocations in
GetFullName have a notable runtime cost.
Here, the allocation is reduced to a single one in every branch.

vepadulano committed:
[core] fix typo in TClingCallFunc::exec

Co-authored-by: Vassil Vassilev <v.g.vassilev@gmail.com>

hageboeck committed:
[tree] Use base class function in TBranchElement::GetFullName().

pcanal committed:
TDataType: add builtins for Double32_t and Float16_t

pcanal committed:
TTreeReaderArray: add support for Double32_t and Float16_t

vgvassilev committed:
Improve the templated Create interface to capture compile-time types.

This interface allows us to set a value and deduce its corresponding type very
efficiently. This is useful when we use cling::Value to model input arguments
for a function call in the JIT (eg. via TClingCallFunc).

That patch essentially makes cling::Value to hold a value and a type that
correctly models the compiled code.

vgvassilev committed:
Make function file static to reduce includes

guitargeek committed:
[RF] Correctly use squared weights in fitTo with `AsymptoticError()`

With the `AsymptoticError()` option, the squared weights were retrieved
as `RooAbsData::weight() * RooAbsData::weight()`. This was wrong for the
RooDataHistCase, because there the sum of squared weights for a single
entry is not the bin content squared. One has to use
`RooAbsData::weightSquared()`, which is correctly overridden in
`RooDataHist`. Now, the reults for an unweighted histogram are again
consistent with the results that you get with `SumW2Error()` and with no
error correction.

linev committed:
Add missing includes in RooSimPdfBuilder.cxx

Appears with -Ddev=on builds

cmsbuild commented:
A new Pull Request was created by  @smuzaffar (Malik Shahzad Muzaffar) for branch cms/master/ef6c047.

@cmsbuild, @smuzaffar, @aandvalenzuela, @iarspider can you please review it and eventually sign? Thanks.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.
cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smuzaffar commented:
please test for CMSSW_12_6_ROOT6_X

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18811e/27153/summary.html
**COMMIT**: d6d609a91c154c0e57e146619e045525e9ae7ec8
**CMSSW**: CMSSW_12_6_ROOT6_X_2022-08-28-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/175/27153/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @fwyzard cms-sw/cmsdist#8046

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18811e/27153/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18811e/27153/git-merge-result


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreServicesDriver had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 16884 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3695708
 - DQMHistoTests: Total failures: 39443
 - DQMHistoTests: Total nulls: 2
 - DQMHistoTests: Total successes: 3656241
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -1.522 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 11834.0 ): -1.518 KiB       SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-08-28-2300+18811e/52436/triggerResults): found differences in 5 / 50 workflows

linev committed:
Mark TObjOptLink methods as override

Prevent warnings when project compiled with -Wsuggest-override option

ferdymercury committed:
[docu] fix typos (#11278)

* [docu] fix typo

* fix typo2

* more typos

guitargeek committed:
[RF] Fix RooDataSet::sumw2() for when weight errors are tracking sumw2

Often, a RooDataSet is used as a binned dataset do do combined fits with
unbinned RooDataSets. Do do the sumw2 error correction right, one would
have to track the sumw2 of the original RooDataHist. A natural way do do
this that was tried by users was to encode the `sumw2` by filling the
RooDataSet in the weight errors. This doesn't work, because the
`RooDataSet::sumw2()` function is hardcoded to return `weight()**2`.

This commit suggests to return `weightError()**2` in the presence of
weight errors to enable the encoding of wum of weight squares in the
weight errors if the user wants to use the error corrections.

Related forum post:
https://root-forum.cern.ch/t/error-for-unbinned-binned-weighted-fits/51336

lmoneta committed:
[tmva] Fix maximum value for indexing numfolds parameter (#11267)

The maximum unsigned int value was used for an integer resulting in a negative integer value and this was causing a bug when accessing the numfold parameter in the CvSplit expression formula.
The bug is present when the parameter "NumFold" is not used in the SPlit expression.
See https://root-forum.cern.ch/t/tmva-crossvalidation-crashes-on-destructor-pyroot/51195

linev committed:
[cling] add override qualifier for ColoredOutput methods

guitargeek committed:
[tree] Add missing override specifiers in tree/ntuple

Sister PR of #11285.

agheata committed:
[geom] Fix for mixture creation based on repeating elements. (#11293)

Neel-Shah-29 committed:
[GSOC][TMVA][SOFIE] Reduce ONNX Operators implemented with the corresponding unit tests (#11258)

* Reduce Operator Initial Code pushed

* Reduce Operator axis attribute added

* Made the required changes according the expected generated code and shape in the Reduce Operator

* Made the required changes to resolve the generated code errors

* Fix generated code for Reduce operator

Fix also correct namespace for reference output in reduce tests

* Reduce Operator code cleaned and fixed the warning for int to size_t when (n==dim-1)

* Reduce ONNX Operator warning fixed related to size_t and int

* Changes related to Reduce Prod done and sum updated to 1 for ReduceProd

Co-authored-by: moneta <lorenzo.moneta@cern.ch>

Neel-Shah-29 committed:
[GSOC][TMVA][SOFIE] Max ONNX Operator implemented with the corresponding unit tests (#11198)

* Max ONNX Operator initial code implemented

* Test added for the Max ONNX Operator

* Implemented the Max Operator for more than 2 inputs as per the changes required

* Max ONNX Operator initial code implemented

* Merge Conflicts resolved

* Three seprate cases added regarding the Input size to optimize the performance of Max Operator

* Multi Broadcasting functionality for same lengths of input and output tensors added in Max ONNX Operator

* Max ONNX Operator initial code implemented

* Test added for the Max ONNX Operator

* Implemented the Max Operator for more than 2 inputs as per the changes required

* Max ONNX Operator initial code implemented

* Merge Conflicts resolved

eguiraud committed:
[NFC][DF] Expand docs for RColumnRegister

eguiraud committed:
[NFC][DF] Fix typo

Co-authored-by: Axel Naumann <Axel.Naumann@cern.ch>

eguiraud committed:
[DF] Rename RDefineAndReaders -> RDefinesWithReaders

This is to signal slightly that it's not a random pair, but the
readers are those of the RDefine.

Similarly for RVariationAndReaders -> RVariationsWithReaders.

Co-authored-by: Axel Naumann <Axel.Naumann@cern.ch>

eguiraud committed:
[DF] Clarify that RDefinesWithReaders::fDefine is never null

eguiraud committed:
[DF] Clarify that RVariationsWithReaders::fVariation is never null

couet committed:
Option X+ and Y+ for TMultiGraph (#11298)

pcanal committed:
cling: Improve support for expression in template re-substitution.

This addresses the issue described in #11259.  In particular it handles the case where the template parameter
is a value.  i.e. in the inner template of
```
__and_<is_constructible<_Rb_tree_iterator<pair<const unsigned int,string> >,const _Rb_tree_iterator<pair<const unsigned int,string> >&>,is_constructible<bool,const bool&> >
```

couet committed:
Implement the "I" option for multigraphs (#11301)

pcanal committed:
Disable gdml tutorial when gdml is off

couet committed:
[skip-ci] improve TGaxis doc (#11308)

vgvassilev committed:
Improve consistency in cling::Value.

The improvements are:
  * We provide getX and setX interfaces instead of returning the address for the
    non const methods. This allows us to be more consistent in terms of
    lifetimes as now users cannot take the address of block of memory which can
    be freed by the cling::Value.
  * We remove the storage types and we rely on the clang::Type which we have in
    the cling::Value.

vgvassilev committed:
Remove unused constructor

vgvassilev committed:
Rename simplisticCast to castAs.

vgvassilev committed:
Implement a rigurous type checking when we call getX and setX.

These interfaces assume we know the type and we should compare if the underlying
type is the one we expect when using the setters and getters. Unfortunately,
this is not the case and we need to further investigate.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

vgvassilev committed:
Improve style and the FIXME

cmsbuild commented:
Pull request #175 was updated.

vgvassilev committed:
Mark m_Storage as a const ref.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

vgvassilev committed:
Improve performance by allowing most of cling::Value to inline.

We introduce an enum which mirrors the type kind and we use it in the getAs and
castAs operations to allow the compiler/interpreter to see all of the functions
and potentially inline them.

This patch brings the performance to similar levels with what we have in the
master.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.

cmsbuild commented:
Pull request #175 was updated.
